### PR TITLE
Simplify high-level streams API

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: quinn/src/connection.rs [hunk 1]
```diff
@@ -597,15 +600,72 @@ impl Write for BiStream {
...
+    fn drop(&mut self) {
         let mut conn = self.conn.lock().unwrap();
-        conn.inner.reset(self.stream, error_code);
+        if conn.error.is_some() {
+            return;
+        }
+        if !self.finished {
+            conn.inner.reset(self.stream, 0);
+        }
         conn.notify();
```
Comment (Ralith) [inline quinn/src/connection.rs]: Should this be inside the `if`?

File: quinn/src/connection.rs [hunk 2]
```diff
@@ -653,76 +713,12 @@ impl Read for BiStream {
...
-        self.poll_finish().map_err(|e| {
-            io::Error::new(
-                io::ErrorKind::ConnectionAborted,
-                format!("connection closed: {}", e),
-            )
-        })
-    }
-}
-
-impl Drop for BiStream {
```
Comment (Ralith) [inline quinn/src/connection.rs]: Missing `impl Drop for RecvStream`?

File: quinn/src/connection.rs [hunk 3]
```diff
@@ -653,76 +713,12 @@ impl Read for BiStream {
             Err(UnknownStream) => Err(ReadError::UnknownStream),
         }
     }
-
-    fn stop(&mut self, error_code: u16) {
```
Comment (Ralith) [inline quinn/src/connection.rs]: Why was this removed?

File: quinn/src/connection.rs [hunk 4]
```diff
@@ -597,15 +600,72 @@ impl Write for BiStream {
...
+impl RecvStream {
+    fn new(conn: ConnectionRef, stream: StreamId) -> Self {
+        Self {
+            conn,
+            stream,
+            recvd: false,
+        }
+    }
+
     fn poll_read_unordered(&mut self) -> Poll<(Bytes, u64), ReadError> {
```
Comment (Ralith) [inline quinn/src/connection.rs]: This and `poll_read` should be `pub`, and be documented by the comments that were on the removed `trait Read`.

File: quinn/src/connection.rs [hunk 5]
```diff
@@ -525,36 +522,42 @@ impl Drop for ConnectionInner {
...
             stream,
             finishing: None,
             finished: false,
-            recvd: false,
         }
     }
-}
 
-impl Write for BiStream {
     fn poll_write(&mut self, buf: &[u8]) -> Poll<usize, WriteError> {
```
Comment (Ralith) [inline quinn/src/connection.rs]: This and `poll_finish` should be `pub`, and documented by the comments that were on the removed `trait Write`.

File: quinn/src/connection.rs [hunk 6]
```diff
@@ -525,36 +522,42 @@ impl Drop for ConnectionInner {
...
     finishing: Option<oneshot::Receiver<Option<ConnectionError>>>,
     finished: bool,
-
-    // Recv only
-    // Whether data reception is complete (due to receiving finish or reset or sending stop)
-    recvd: bool,
 }
 
-impl BiStream {
+impl SendStream {
```
Comment (Ralith) [inline quinn/src/connection.rs]: Missing `pub fn reset(&mut self, error_code: u16)` and associated documentation

File: quinn/src/connection.rs [hunk 7]
```diff
@@ -742,71 +738,14 @@ impl io::Read for BiStream {
...
-
 impl AsyncRead for RecvStream {
     unsafe fn prepare_uninitialized_buffer(&self, _: &mut [u8]) -> bool {
         false
     }
 }
 
 /// Uses unordered reads to be more efficient than using `AsyncRead` would allow
-pub fn read_to_end<T: Read>(stream: T, size_limit: usize) -> ReadToEnd<T> {
+pub fn read_to_end(stream: RecvStream, size_limit: usize) -> ReadToEnd {
```
Comment (Ralith) [inline quinn/src/connection.rs]: Would it make sense to move this to a method on `RecvStream`, now that it doesn't need to be generic?

Comment (djc) [reply quinn/src/connection.rs]: Yup.

File: quinn/src/connection.rs (hunk 2 above)
Comment (djc) [reply quinn/src/connection.rs]: Yeah...

File: quinn/src/connection.rs (hunk 3 above)
Comment (djc) [reply quinn/src/connection.rs]: Because it was not `pub`, the compiler flagged it as unused (similar to `SendStream::reset()`). Added both back as `pub` methods.

File: quinn/src/connection.rs (hunk 4 above)
Comment (djc) [reply quinn/src/connection.rs]: Fixed.

File: quinn/src/connection.rs (hunk 1 above)
Comment (djc) [reply quinn/src/connection.rs]: Looks sensible!

File: quinn/src/connection.rs (hunk 5 above)
Comment (djc) [reply quinn/src/connection.rs]: Yup.

File: quinn/src/connection.rs (hunk 6 above)
Comment (djc) [reply quinn/src/connection.rs]: Added it back.

**Comment by djc** [comment]
Given the way tests ended up unbundling `BiStream`s anyway, I figured I'd pre-unbundle them, and I like how it ended up. See what you think.

File: quinn/src/connection.rs [hunk 8]
```diff
@@ -609,14 +610,80 @@ impl Write for BiStream {
...
+    }
+}
+
+impl AsyncWrite for SendStream {
+    fn shutdown(&mut self) -> Poll<(), io::Error> {
+        self.poll_finish().map_err(|e| {
+            io::Error::new(
+                io::ErrorKind::ConnectionAborted,
+                format!("connection closed: {}", e),
+            )
```
Comment (Ralith) [inline quinn/src/connection.rs]: `e.into()`?

File: quinn/src/connection.rs (hunk 4 above)
Comment (Ralith) [reply quinn/src/connection.rs]: Looks like it's still missing `pub`.

File: quinn/src/connection.rs [hunk 9]
```diff
@@ -641,6 +708,17 @@ impl Read for BiStream {
...
+    /// Returns the number of bytes read into `buf` on success.
+    ///
+    /// Applications involving bulk data transfer should consider using unordered reads for
+    /// improved performance.
+    ///
+    /// # Panics
+    /// - If called after `poll_read_unordered` was called on the same stream.
+    ///   This is forbidden because an unordered read could consume a segment of data from a
+    ///   location other than the start of the receive buffer, making it impossible for future
     fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, ReadError> {
```
Comment (Ralith) [inline quinn/src/connection.rs]: Missing `pub`.

File: quinn/src/connection.rs (hunk 8 above)
Comment (djc) [reply quinn/src/connection.rs]: Yup.

File: quinn/src/connection.rs (hunk 9 above)
Comment (djc) [reply quinn/src/connection.rs]: Oops, sorry I missed that.
